### PR TITLE
"AM" 9: handle "AppMan" programs without switching to "AppMan Mode"

### PR DESCRIPTION
**Pull request by ivan-hc**
see https://github.com/ivan-hc/AM/discussions/1034

As requested, this change aims to allow "AM" installing (option `-i`) and listing installed apps (option `-f`) also as AppMan user, without having to switch to "AppMan Mode" (option `--user`) and go back (`--system).

### What to do

- [x] -a can detect f an app is installed or not
- [x] -b for backups of apps that may be needed for both AM and AppMan
- [x] -C to create potable .config directories
- [x] -e it is difficult, since it needs more arguments
- [x] -f done! Shows apps installed from both.
- [x] -H to create portable .home directories
- [x] -i done, to install apps.
- [x] -ia done, to install appimages.
- [x] -o to restore backups, so same ad -b
- [x] -r and -R to remove apps
- [x] -u done, AM can update also apps for AppMan now... I need to rework the way the apps are updated btw
- [x] --icons to get icons from installed apps, to use themes
- [x] lock to lock version and updates
- [x] unlock to undo "lock"
- [x] nolibfuse to convert old appimages to new runtime
- [x] sandbox to sandbox appimages, here I need the help of @Samueru-sama 
- [x] --disable-sandbox
- [x] --force-latest to downgrade an app quickly if it is a dev build or an alpha
- [x] --rollback, to downgrade the apps from a list
- [x] -c to clean unneeded files and the cache
- [x] -l done
- [x] -s to check for installation scripts changes among the installed apps with this database
- [x] -h added info about the usage of --user
- [x] --user suggested as a flag instead of using it as an option
- [x] -d , suboption --convert

### What not to do
- [x] -q dont needs changes
- [x] -t dont needs changes
- [x] -v dont needs changes
- [x] --launcher dont needs changes
- [x] --devmode-enable/disable dont needs changes
- [x] --system dont needs changes
- [x] apikey dont needs changes, key is shared
- [x] newrepo is shared, dont needs changes

### CODE REVIEW

**Comment by ivan-hc** [comment]
## Option `-i` or `files`
I have just tested this for `-f` to allow "AM" see installed apps also in AppMan, while "AppMan" or 2AppMan Mode" are not allowed

https://github.com/user-attachments/assets/1311da9f-91fa-42e4-a405-ace2ae3fcbca

the trick is to check if the file "appman-config" exists

https://github.com/user-attachments/assets/c4e9436e-65c3-4a19-ba6b-942cacb3228e

the responsible of this is a new function I added in "database.am"
```
_files_check_appman_from_am() {
	[ "$AMCLI" = am ] && [ -f "$APPMANCONFIG"/appman-config ] && _appman && echo "$DIVIDING_LINE" && APPMAN_ON=1
}
```
that is applied in the option like this
```
	'-f'|'files')
		if [ "$2" = "--less" ]; then
			_files_show_only_number
			_files_check_appman_from_am && _files_show_only_number
		elif [ "$2" = "--byname" ]; then
			_files_sort_by_name
			_files_check_appman_from_am && _files_sort_by_name
			_betatester_message_on
		else
			_files_sort_by_size
			_files_check_appman_from_am && _files_sort_by_size
			_betatester_message_on
		fi
		;;
```
plus some corrections to the messages
```
_files_header() {
	echo ""
	APPSNUMB=$(find "$APPSPATH" -type f -name 'remove' -exec grep -L 'usr/local/lib' {} \; 2>/dev/null | sed -r 's|/[^/]+$||' | sort | uniq | wc -l)
	if [ "$APPSNUMB" = 1 ]; then
		APPSMESSAGE="YOU HAVE INSTALLED $APPSNUMB PROGRAM"
	else
		APPSMESSAGE="YOU HAVE INSTALLED $APPSNUMB PROGRAMS"
	fi
	if grep -q 'usr/local/lib' "$APPSPATH"/*/remove 2> /dev/null; then
		LIBNUMB=$(grep -l "usr/local/lib" "$APPSPATH"/*/remove | wc -l)
		if [ "$LIBNUMB" = 1 ]; then
			LIBSMESSAGE="AND $LIBNUMB LIBRARY"
		else
			LIBSMESSAGE="AND $LIBNUMB LIBRARIES"
		fi
		[ "$APPMAN_ON" = 1 ] && echo "- $APPSMESSAGE $LIBSMESSAGE IN \"APPMAN MODE\", LOCALLY" || echo "- $APPSMESSAGE $LIBSMESSAGE MANAGED BY \"$AMCLIUPPER\""
	else
		[ "$APPMAN_ON" = 1 ] && echo "- $APPSMESSAGE IN \"APPMAN MODE\", LOCALLY" || echo "- $APPSMESSAGE MANAGED BY \"$AMCLIUPPER\""
	fi
	echo ""
}
```
to test all this (and join the beta-testing) run `am --devmode-enable` and `am -s`

This change does not affects the "pure AppMan" installation.

**Comment by ivan-hc** [comment]
in the last commit I moved the function from database.am to APP-MANAGER, so we can re-use it in other options if needed.

Now the function in APP-MANAGER is this
```
function _check_appman_from_am() {
	[ "$AMCLI" = am ] && [ -f "$APPMANCONFIG"/appman-config ] && _appman && echo "$DIVIDING_LINE" && APPMAN_ON=1
}
```
I can now test it in `-i`

**Comment by ivan-hc** [comment]
## Option `-i` or `install`
Now the option can be used as suboption in `-i`

https://github.com/user-attachments/assets/28da0eb9-ed94-4340-8cf6-e699dd3dfd03

I hink I need improvements for this, @Samueru-sama any idea?

**Comment by ivan-hc** [comment]
## Option `-u` or `update`
I also managed to made "AM" updating apps installed with AppMan (sorry for the slow updates in the video, I reached GH API calls and I'm now using `torsocks` in support):

https://github.com/user-attachments/assets/593f9d11-6ef6-49af-9dde-1a5f2a897c81

**Comment by ivan-hc** [comment]
What I don't understand is why in both AppMan Mode and this new functions, the updates of apps installed locally is not shown.

**Comment by ivan-hc** [comment]
> What I don't understand is why in both AppMan Mode and this new functions, the updates of apps installed locally is not shown.

uhm... it seems that the issue is related to "$AMCACHEDIR", in "AppMan Mode" it is stil ~/.cache/am

Anyway, in AppMan Mode I'm still able to see `-f` for AppMan. It is `-u` that can't see the versions to compare them.

**Comment by ivan-hc** [comment]
https://github.com/ivan-hc/AM/pull/1036/commits/2438095efa5621e5ec80d1c8a23e04bd6521fcd9 in this commit I made flags `--debug`, `--force-latest` and `--user` free and interchangeable, and now also option `-ia` or `install-appimage` supports the --user flag

**Comment by ivan-hc** [comment]
I need a function to quickly switch between AM and AppMan Mode for the tasks that still need to be perfected

**Comment by ivan-hc** [comment]
Last commit adds this support to -a, no --user flag is needed btw

**Comment by ivan-hc** [comment]
My latest commit allows AM to handle all options from the sandboxes.am module for apps installed via AppMan

@Samueru-sama take a look at it https://github.com/ivan-hc/AM/pull/1036/commits/eed8c25af83395c985833be8b07b6a163b9a1e6f

**Comment by ivan-hc** [comment]
And now also --force-latest can handle AppMan's apps from AM

**Comment by ivan-hc** [comment]
This time options `-b`/`backup` and `-o`/`overwrite` can be used with AM to manage apps installed via AppMan.

**Comment by ivan-hc** [comment]
Now also the other rootless options `lock`, `unlock`, `nolibfuse` and `downgrade`/`--rollback` can manage AppMan's apps via AM

Now only `-R` and `-r` are left... and here I have an issue, I have tested a `-R` in bulk by mixing the items, and the one installed system-wide results removed in the terminal, but it is not removed.

**Comment by ivan-hc** [comment]
Also option `icons` is ready

**Comment by ivan-hc** [comment]
I'm stuck on `-R` and `-r` to allow AM removing local apps near the system apps.

@Samueru-sama since you did the last changes, can you help me?

**Comment by ivan-hc** [comment]
> I'm stuck on `-R` and `-r` to allow AM removing local apps near the system apps.

I have finally solved

**Comment by ivan-hc** [comment]
Since APP-MANAGER and all modules are involved in this really huge change... this deserve to be a main point release

**Comment by ivan-hc** [comment]
I want to take some time to let you test the release, run the command `am --devmode-enable` or `appman --devmode-enable` and then update with `-s`.

AppMan users should not notice any difference, but "AM" itself has changed. A lot! It can also do the job of AppMan, if its configuration file is present, as I showed in the first videos.

Install apps, update, remove, use sandbox, snapshot or whatever you want to do... even install third-party AppImages (option `-e`). The important thing is that you report any bugs.

I think I won't publish this release before Sunday.

Thanks for the contribution.

File: modules/management.am [hunk 1]
```diff
@@ -326,34 +333,43 @@ function _nolibfuse() {
...
 # REMOVE
+function _detect_appman_apps() {
+	APPMAN_APPSPATH=$(<"$APPMANCONFIG/appman-config")
+	if ! echo "$APPMAN_APPSPATH" | grep -q "^/"; then
+		APPMAN_APPSPATH="$HOME/$(<"$APPMANCONFIG/appman-config")"
+	fi
+}
+
+function _detect_path_to_remove() {
+	[ -d "$APPMAN_APPSPATH"/"$arg" ] && RMPATH="1" || [ -d "$APPSPATH"/"$arg" ] && RMPATH="1" || RMPATH=""
```
Comment (Samueru-sama) [inline modules/management.am:344]: you might want to turn this into a if elif else statement

**Comment by contrarybaton60** [comment]
no more "AppMan" perhaps? I suppose it's easier to integrate AppMan's way to install programs in AM than keep them separate, right?

**Comment by contrarybaton60** [comment]
image

**Comment by ivan-hc** [comment]
> no more "AppMan" perhaps? I suppose it's easier to integrate AppMan's way to install programs in AM than keep them separate, right?

AM needs to be installed as root and is owned by the one that have installed it. In this sense, AppMan is more flexible. So nope, they will keep live side by side. Don't worry.

File: modules/management.am (hunk 1 above)
Comment (ivan-hc) [reply modules/management.am:344]: well, it was an `if`/`fi` statement as I have shown on Discord... but before publishing my code I have tested also this form and worked, so I have done a one-line command to save space

Comment (Samueru-sama) [reply modules/management.am:344]: It also makes it harder to read, and also the behavior is actually slightly different than that of a if else statement. see this example: 

image

Comment (ivan-hc) [reply modules/management.am]: done https://github.com/ivan-hc/AM/pull/1036/commits/c0253f8fc00303d0a4f84503026dd0a470ad41d7

**Comment by ivan-hc** [comment]
I noticed a series of bugs that now have been fixed.

In short, if there was no "appman-config" file, many options refused to start or looped.

Now everything is fixed. At most I would like to fix the welcome message in "AppMan", now startable if there is a "--user" flag, currently used for the "-i", "-ia" and "-e" installation options.

**Comment by ivan-hc** [comment]
Thanks to a new variableI named "APPSPATH_ORIGIN" I'm now able to fix wrong loops.

In brief, at start of each module I record the original value of "APPSPATH", so if the variale to set "APPSPATH" from AppMan in AM keeps its value for the other arguments, it unsets the value and restores it to the original one.

**Comment by ivan-hc** [comment]
Since I have been testing the bugs brought by the new change over and over again in the last 2 hours, fixing them... it is wise to keep this release in testing.

I look forward to your feedback.

Remember: `am --devmode-enable` and `am -s` to enter the developer mode

**Comment by ivan-hc** [comment]
Whit this commit https://github.com/ivan-hc/AM/pull/1036/commits/b363912f38e3359fff8b0d62b536a4068200224d, AM will show AppMan apps only if they exist, else it will run normally, by showing only system apps.

Whit this commit instead https://github.com/ivan-hc/AM/pull/1036/commits/be307dfefd97239126dd34b9255752cca4017bbe, AM will remove the apps asking root password only if the app selected (no matter where you place it in a bulk remove command with multiple arguments) has a read-only directory.

**Comment by Twig6943** [comment]
Release date for this pr? (really excited)

**Comment by ivan-hc** [comment]
> Release date for this pr? (really excited)

to quote a famous Italian painter, Michelangelo Buonarroti, "when I've finished"

But you can test it in preview

> Remember: `am --devmode-enable` and `am -s` to enter the developer mode

**Comment by ivan-hc** [comment]
I'm currently investigating any issues that might be there.

I'd be more interested in hearing from AppMan users if they run into any issues, in theory they shouldn't notice any differences.

It's an update that will allow AM to do everything AppMan does, without changing the CLI.

**Comment by ivan-hc** [comment]
In the first comment https://github.com/ivan-hc/AM/pull/1036#issue-2616768899 you all can see the checkbox with all options that have been changed.

The ones not checked are `-c` and `-l` because I have doubts doing changes for them.

Under "**What not to do**" there are listed options that don't require changes as well, but are listed only to be listed near the others. They are ready "by default".

The only thing, the `-h` or `help` option, is changed like this, talking about `--user` as a "flag" in the installation-related options:

Istantanea_2024-10-29_19-50-45 png

**Comment by ivan-hc** [comment]
Maybe... is it the case of drop options `--system`, `appman` and `--user` to completelly switch to `--user` as a flag?

**Comment by ivan-hc** [comment]
> Maybe... is it the case of drop options `--system`, `appman` and `--user` to completelly switch to `--user` as a flag?

Maybe not, since `-l`, `-c` or the ApMan Mode itself is still used or may be useful for many.

Maybe I should add a message with confirm by suggesting the use of `--user` as a flag instead of switching to AppMan Mode.

**Comment by ivan-hc** [comment]
> Maybe I should add a message with confirm by suggesting the use of `--user` as a flag instead of switching to AppMan Mode.

done

Istantanea_2024-10-29_20-59-43 png

**Comment by ivan-hc** [comment]
At this point, also the README needs a rework.... :face_with_head_bandage: what a mess

**Comment by Twig6943** [comment]
@ivan-hc  speaking of the readme please shorten it (create more .md files instead of putting everything to readme.md) (my browser kinda lags whenever I visit the main page)

**Comment by ivan-hc** [comment]
I need to rework the apps update approach now

**Comment by ivan-hc** [comment]
Now AM can use `-c` on AppMan files

**Comment by ivan-hc** [comment]
I rewrote the way the apps to be updated are listed and updated

**Comment by ivan-hc** [comment]
with the last commit, I've finished with APP-MANAGER

I think I should fix something else in modules depending from some new functions in the main CLI

**Comment by ivan-hc** [comment]
Last module left to finalize.

**Comment by ivan-hc** [comment]
Now everything should be ready.

I can start fixing the opened issues, I'm sorry for the long waiting.

Again, if you want help me check for bugs and malfunctions, run `am --devmode-enable` and `am -s` to enter the development mode and use this build in preview.

To exit the developer mode, use `am --devmode-disable`. same options work also for AppMan users, as always.

See you later.

**Comment by ivan-hc** [comment]
> @ivan-hc speaking of the readme please shorten it (create more .md files instead of putting everything to readme.md) (my browser kinda lags whenever I visit the main page)

I think I'll start working on the readme and the documentation tomorrow.

**Comment by ivan-hc** [comment]
For all of you who are testing this preview release, please leave me some feedback. Tell me what options you tried and if you noticed anything different.

The only option that has changed is `--user`, which can be used as a "flag" after the installation options `-i` (normal installation), `-ia` (install only AppImages) and `-e` (install AppImages from github that are not in the database, or are there but you want to give them a different name).

Istantanea_2024-10-31_04-25-59 png

I remind you that to enter the developer mode you need to run the following commands:
```
am --devmode-enable
am -s
```
to exit it, use instead
```
am --devmode-disable
```
and then `am -s` to restore "AM" to the "stable" release, in "main". This is the "dev" branch instead.

If you already have apps for AppMan, you should see them listed in `am -f` or when you update them with `am -u` and `am -u --apps`.

The only option left to change is `-l` or `list`, it still only shows system-wide apps. I will eventually do something different for this option. I have yet to decide.

Please help me get this version out as soon as possible.

@Samueru-sama @Sush-ruta @zen0bit @nazdridoy @Twig6943 @DLopezJr @MANICX100 @Wesley-AlvesRolim @esdrasjnr @stasadev @xz-dev @marco-calautti @contrarybaton60 @fiftydinar

**Comment by Twig6943** [comment]
Ill try to test it asap

**Comment by ivan-hc** [comment]
There are always details to correct

Istantanea_2024-10-31_06-59-11 png

**Comment by Sush-ruta** [comment]
I have been using it for a day or so with no obvious problems. The combined update thing was a surprise (I did not read update notes lol). One thing: `am -l` shows the number of installed apps which are done with `am` but the list shows all apps (including those installed with `appman`).

Screenshot at 15-03-49

**Comment by ivan-hc** [comment]
> One thing: `am -l` shows the number of installed apps which are done with `am` but the list shows all apps (including those installed with `appman`).

I didn't notice that... it's the only option that needs tweaking and I haven't touched it yet. Thanks for the heads up.

**Comment by ivan-hc** [comment]
> One thing: `am -l` shows the number of installed apps which are done with `am` but the list shows all apps (including those installed with `appman`).

@Sush-ruta solved

https://github.com/user-attachments/assets/1e44934f-21af-4dca-a286-917c1de09cdb

**Comment by ivan-hc** [comment]
Creating a hidden `.am-installer` directory should be a feature of apps that have an `AM-updater` file, since they are the only ones that depend solely on AM to receive updates. So Firefox or Thunderbird do not need it... and neither "AM" (directory `/opt/am`).

Previously the function that created this directory took as "arguments" only the directories containing an AM-updater, but since the variable $ARG has changed globally to AM, I had to add this exception to the function, precisely excluding the creation of this directory in the aforementioned cases.

Here's what the latest commit means https://github.com/ivan-hc/AM/pull/1036/commits/8b106cc7b03c6b6958d8478f423c858ec28751a8

**Comment by ivan-hc** [comment]
Also `-d --convert` was a unique feature of AppMan... not anymore https://github.com/ivan-hc/AM/pull/1036/commits/130a1baab010c1b6a37ef0839fa0227178b4f892

It seems that there were many more options to change.

**Comment by ivan-hc** [comment]
Regarding the documentation, I opened a new branch to modify the README 

https://github.com/ivan-hc/AM/tree/update-readme

I opened a PR as well for this https://github.com/ivan-hc/AM/pull/1068

what i removed is only temporary. i will rewrite the bits and pieces little by little... maybe i was planning to add everything directly in "Usage" instead of creating secondary subpages... not sure.

**Comment by ivan-hc** [comment]
The README is updated.

The documentation has been split into multiple markdown files reachable from the main index.

You can find the documentation files in the "docs" directory.

**Comment by ivan-hc** [comment]
Ready?

**Pull request merged**